server/modules/surveyRdb/repository/dataView/read.js

```diff
@@ -12,63 +12,207 @@ import * as Expression from '../../../../../core/expressionParser/expression'
+  return ColumnNodeDef.getColName(nodeDefParent)
+}
+
+/**
+ * Determines the select fields to be used in the query.
+ *
+ * @param {!object} params - The query parameters.
+ * @param {!Survey} [params.survey] - The survey.
+ * @param {!NodeDef} [params.nodeDef] - The node def associated to the view to select.
+ * @param {Array} [params.nodeDefCols=null] - The node defs associated to the selected columns.
+ * @param {boolean} [params.columnNodeDefs=false] - Whether to select only columnNodes.
+ * @param {boolean} [params.editMode=false] - Whether to fetch row ready to be edited (fetches nodes and records).
+ *
+ * @returns {string[]} - An array of fields to be used in the select query.
+ */
+const _getSelectFields = (params) => {
+  const { survey, nodeDef, columnNodeDefs, nodeDefCols, editMode } = params
+  const viewDataNodeDef = new ViewDataNodeDef(survey, nodeDef)
+
+  if (nodeDefCols) {
+    const selectFields = [
+      viewDataNodeDef.columnRecordUuid,
+      viewDataNodeDef.columnUuid,
+      // selected node def columns
+      ...nodeDefCols.map((nodeDefCol) => `${viewDataNodeDef.alias}.${ColumnNodeDef.getColName(nodeDefCol)}`),
+    ]
+    // Add ancestor uuid columns
+    Survey.visitAncestorsAndSelf(nodeDef, (nodeDefAncestor) => {
+      if (!(NodeDef.isVirtual(nodeDefAncestor) || NodeDef.isRoot(nodeDefAncestor))) {
+        const columnUuid = `${viewDataNodeDef.alias}.${ColumnNodeDef.getColName(nodeDefAncestor)}`
+        selectFields.push(columnUuid)
+      }
+    })(survey)
+
+    if (editMode) {
+      selectFields.push(
+        ...[
+          // Node (every node is transformed into json in a column named with the nodeDefUuid)
+          ...nodeDefCols.map((nodeDefCol, idx) => `row_to_json(n${idx + 1}.*) AS "${NodeDef.getUuid(nodeDefCol)}"`),
+          // Record table fields
+          'row_to_json(r.*) AS record',
+        ]
+      )
+    }
+    return selectFields
+  }
+  if (columnNodeDefs) {
+    return [viewDataNodeDef.columnRecordUuid, ...viewDataNodeDef.columnNodeDefNamesRead]
+  }
+  return '*'
+}
+
+/**
+ * Generates the from clause to be used in the select query.
+ *
+ * @param {!object} params - The query parameters.
+ * @param {!Survey} [params.survey] - The survey.
+ * @param {!NodeDef} [params.nodeDef] - The node def associated to the view to select.
+ * @param {Array} [params.nodeDefCols=null] - The node defs associated to the selected columns.
+ * @param {boolean} [params.columnNodeDefs=false] - Whether to select only columnNodes.
+ * @param {boolean} [params.editMode=false] - Whether to fetch row ready to be edited (fetches nodes and records).
+ *
+ * @returns {string} - The from clause.
+ */
+const _getFromClause = (params) => {
+  const { survey, nodeDef, editMode, nodeDefCols } = params
+  const viewDataNodeDef = new ViewDataNodeDef(survey, nodeDef)
+  const { surveyId } = viewDataNodeDef
+
+  const fromTables = [viewDataNodeDef.nameAliased]
+  if (editMode) {
+    fromTables.push(
+      ...[
```

Comment:
do you need ...[...map ? would just ...map be enough ?

Reply:
right